### DISCUSSION TITLE
How to disable css/sass/... processing?

### DISCUSSION THREAD

**Discussion by eduardweigandt0770** [vitest-dev/vitest#498] [Q&A] [answered]
In jest you would have a transformer/mock for all imports with styling resources. How can I disable such processing with vitest?

**Comment by sheremet-va** [comment]
Do you need to tho? Vite can handle them, Jest cant

**Reply by eduardweigandt0770** [reply]
For unit tests I do not need styling and in my case I just don't want to wait until the whole Vuetify stylsheets are compiled.

**Reply by patak-cat** [reply]
You could add an alias to an empty stylesheet, or add an inline plugin that returns an empty virtual stylesheet for these files

**Reply by sheremet-va** [reply]
You can do something like this:

```ts
// vite.config.ts
{
   resolve: {
      alias: [
         ...(mode === 'test' && [{ find: /\.css$/, replacement: 'path to empty styleshit' }])
      ]
   }
}
```

**Reply by sheremet-va** [reply]
I think Vitest can provide a shortcut to this inside `test` property

**Reply by eduardweigandt0770** [reply]
With this config I get a load error with wrong paths, because I match and replace only the extension. So should I write a plugin, because aliases only find/replace  parts?
```ts
// vite.config.ts
{
   resolve: {
      alias: [
        { find:'@', replacement: new URL('./src/', import.meta.url).pathname },
        { find: /\.(scss|sass|css)$/, replacement: new URL('./tests/empty.css', import.meta.url).pathname },
      ],
   }
}
```

and the error:
`
 ❯ tests/unit/components/HelloWorld.spec.js (0)

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  tests/unit/components/HelloWorld.spec.js [ tests/unit/components/HelloWorld.spec.js ]
Error: failed to load ../../../src/components/VApp/VApp/home/user/Workspaces/tmp/migration/tests/empty.css
 ❯ async node_modules/.pnpm/vuetify@2.6.2_vue@2.6.14/node_modules/vuetify/lib/components/VApp/VApp.js:2:31
      1| // Styles
      2| import "../../../src/components/VApp/VApp.sass"; // Mixins
       |                               ^
      3| 
      4| import Themeable from '../../mixins/themeable'; // Utilities
 ❯ async node_modules/.pnpm/vuetify@2.6.2_vue@2.6.14/node_modules/vuetify/lib/components/VApp/index.js:1:310
 ❯ async node_modules/.pnpm/vuetify@2.6.2_vue@2.6.14/node_modules/vuetify/lib/components/index.js:1:310
`

**Reply by eduardweigandt0770** [reply]
With this regex the find/replace works:

```ts
  resolve: {
    alias: [
      { find:'@', replacement: new URL('./src/', import.meta.url).pathname },
      { find: /.*\.(?:scss|sass|css)$/, replacement: new URL('./tests/empty.css', import.meta.url).pathname },
    ],
  }
```

**Reply by Akshay73c** [reply]
> I think Vitest can provide a shortcut to this inside `test` property

How?? Will you please explain ?

**Reply by sheremet-va** [reply]
> > I think Vitest can provide a shortcut to this inside `test` property
> 
> How?? Will you please explain ?

Vitest has `test.alias`.

**Comment by antfu** [comment, marked as answer]
An inline plugin might be better?

```ts
plugins: [
  {
    enfore: 'pre',
    transform(code, id) {
      if (id.endsWith('.css'))
        return ''
    }
  }
]
```

**Reply by eduardweigandt0770** [reply]
Thank you, it also works with this plugin:
```ts
        {
            enforce: 'pre',
            transform(code, id) {
                if (/\.(css|sass|scss)$/.test(id)) return { code: '' };
            }
        }
```

**Reply by CyberAP** [reply]
@antfu would it possible to make it a default for vitest? I think there's little to no reason to process styles for unit testing by default. I might be wrong of course.

**Reply by kingyue737** [reply]
How can I add this plugin only during test? I want to keep all the other plugins used in dev/production mode during test.

**Reply by kingyue737** [reply]
I found that falsy plugin is ignored by Vite. The code below works

```
plugins:[
  process.env.VITEST
     ? {
           enforce: 'pre',
           transform(code, id) {
              if (/\.(css|sass|scss)$/.test(id)) return { code: '' }
           },
        }
      : undefined,
]
```

**Reply by ascott18** [reply]
Would love to see this as a default. Adding this sped up the warmup time of my tests from 16s to 6s.

**Reply by patak-cat** [reply]
See PR from @sheremet-va:
- https://github.com/vitest-dev/vitest/pull/1467

**Reply by kingyue737** [reply]
The new feature does not seem to work. In v0.15.0, if I remove the inline plugin indicated above, test time increases 20 seconds.

with inline plugin:

![image](https://user-images.githubusercontent.com/40021217/173767361-665d7bdf-70ed-466a-8349-e0df28279324.png)


remove inline plugin and set `test.css` `false` in `vite.config` (although it should not process css by default):

![image](https://user-images.githubusercontent.com/40021217/173766965-62098148-55a6-4366-bcbc-60a2386c898f.png)

**Reply by sheremet-va** [reply]
While debugging I see that we return empty css. Do you have a reproduction?

**Reply by kingyue737** [reply]
@sheremet-va  https://stackblitz.com/edit/vitest-dev-vitest-vaeyiy

The deprecation warnings of `sass` shown in terminal indicates that `vitest` still process sass code of `Vuetify`

**Reply by sheremet-va** [reply]
> @sheremet-va https://stackblitz.com/edit/vitest-dev-vitest-vaeyiy
> 
> The deprecation warnings of `sass` shown in terminal indicates that `vitest` still process sass code of `Vuetify`

Can you provide a reproduction with working tests and not working css processing? If I add `vuetify` to `deps.inline` css is not processed for me.

**Reply by sheremet-va** [reply]
Ah, I see. If I replace `return ''` to `replace { code: '' }` I don't see deprecation errors. Will create a PR.

**Reply by sheremet-va** [reply]
Fix released in 0.15.1

**Reply by patak-cat** [reply]
Uh, good catch. Is this a bug in Vite? I would expect your original code should give the same result

**Reply by sheremet-va** [reply]
> Uh, good catch. Is this a bug in Vite? I would expect your original code should give the same result

I think it might be. Maybe `(!result)` check instead of `result != null`?
